File: java/com/blind75/arrays/easy/TwoSum.java
package com.blind75.arrays.easy;

import com.blind75.helper.Task;

import java.util.HashMap;

public class TwoSum extends Task {

    /**
     * O(n^2) time complexity - every for loop add time complexity.
     *
     * @param nums
     * @param target
     * @return
     */
    public int[] bruteForce(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    return new int[]{i, j};
                }
            }
        }
        return new int[0];
    }

    /**
     * O(n) complexity in time and space
     * and faster.
     *
     * @param nums
     * @param target
     * @return
     */
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i])) {
                return new int[]{map.get(nums[i]), i};
            }
            map.put(target - nums[i], i);
        }
        return new int[0];
    }

    public static void main(String[] args) {
        new TwoSum().run();
    }

    @Override
    protected void run() {
        int[] nums = {2, 7, 12, 23};
        int target = 9;
        printArray(bruteForce(nums, target), "bruteforce");
        printArray(twoSum(nums, target), "optimized");
    }
}

File: vijay/python/blind75/twosum.py
"""
Return a list containing the index pair
of a list that contains the sum
of a target
"""
from typing import List

# O(N^2) Time Complexity
# O(N) Space Complexity
def bruteForce(nums:List[int], target:int) -> List[int]:
    for i in range(len(nums)):
        for j in range(i+1, len(nums)):
            if target - nums[i] == nums[j]:
                return [i, j]
    return []

# O(N) Time Complexity
# O(N) Space Complexity
def optimized(nums:List[int], target:int) -> List[int]:
    lookup = {}
    for i in range(len(nums)):
        print(lookup.get(target - nums[i]))
        if lookup.get(target - nums[i]) is not None: 
            print('true')
            return [lookup.get(target - nums[i]), i]
        lookup[nums[i]] = i

    
nums = [2,7,11,15]
target = 9

# print(bruteForce(nums, target))
print(optimized(nums, target))